[MM-26721] show channels across teams in channel switcher modal

AshishDhama opened the pull request:
#### Summary
When the user types in the channel switcher, match results across all teams on the server.

If the user has more than 1 team:

The team name would be on the right

The mention badge would be beside the name of the Channel.

If the user has only 1 team:
we have decided to keep the existing layout i.e no changes in the ``channel switcher`` list view.

#### Ticket Link
Fixes [https://mattermost.atlassian.net/browse/MM-26721](https://mattermost.atlassian.net/browse/MM-26721)

#### Related Pull Requests
- Has server changes [https://github.com/mattermost/mattermost-server/pull/18468](https://github.com/mattermost/mattermost-server/pull/18468)

#### Screenshots
|  before  |  after  |
|----|----|
| <img width="623" alt="Screenshot 2021-10-15 at 8 31 41 PM" src="https://user-images.githubusercontent.com/16203333/137509301-e6b77a5f-849a-4274-aa86-b2e054123f4d.png"> |  <img width="622" alt="Screenshot 2021-10-15 at 8 27 04 PM" src="https://user-images.githubusercontent.com/16203333/137509372-9ff12310-4967-4015-8ec1-071085457c7b.png"> |


#### Release Note
```release-note
* API: modification in existing API endpoints to meet the Requirements (added new params).
* UI: Behavioral changes in channel switcher if user part of more than 1 team UI will show up with team name.
* Redux: Modification in a couple of redux selectors and added few selectors for selecting all the channels 
```

AshishDhama commented:
@agnivade adding you for visibility

angeloskyratzakos commented:
/update-branch

agnivade commented:
@jgilliam17 - Just a quick note that this might need some extra testing which might not be possible via spinwicks. This needs testing with Elasticsearch and Bleve as well. ES is likely not available with spinwicks.

For the basic feature:
- Verify that public and private channels from all teams are visible in the switcher, and is navigable.
- Verify that channels (from all teams) of which the user is not a member of, is also searchable in the switcher.

For ES and Bleve, you would need to disable database search, and then run the indexing job, and enable search and autocomplete.

Please feel free to let me know if you have any questions. :)

AshishDhama commented:
/update-branch

esethna commented:
Hey @AshishDhama just checking on this and what are next steps?

AshishDhama commented:
@esethna yeah will update this with changes suggested next week, then I guess it would be good to merge

Willyfrog commented:
@AshishDhama anything blocking this?

AshishDhama commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

esethna commented:
Hey @AshishDhama, can you provide an update on where this PR is at?

AshishDhama commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -199,16 +209,22 @@ class SwitchChannelSuggestion extends Suggestion {
             >
                 {icon}
-                <div className='suggestion-list__ellipsis'>
+                <div className={classNames('suggestion-list__ellipsis', {'suggestion-list__flex': !isPartOfOnlyOneTeam})}>
```
Make sense but I am planning to break this component into smaller components, right now I have just modified it according to the needs for this feature, will take care of this when we will be breaking this into small components to reduce the complexity.

AshishDhama commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -655,13 +674,13 @@ export default class SwitchChannelProvider extends Provider {
         return channelList;
     }
 
-    async fetchChannels(resultsCallback, size = 20) {
```
Yeah as the endpoint we are calling for all team channels is handling that on backend

AshishDhama commented:
> Hey @AshishDhama, can you provide an update on where this PR is at?

@esethna I think we are ready to merge, have implemented all the changes suggested by @hmhealey except the ones which were increasing the scope of this ticket.

AshishDhama commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/102943

esethna commented:
@Willyfrog @AshishDhama great, can we rally reviews to get this PR merged? It's a massive quality of life enhancement for multi-team users

esethna commented:
Ping @hmhealey @koox00 @jgilliam17 for reviews

AshishDhama commented:
/update-branch

AshishDhama commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/103082

AshishDhama commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/103176

hmhealey commented:
Are there still changes outstanding on this? I don't see anything in the web app code about loading channel members for other teams, and we'd need that to show unreads for channels in teams that haven't been loaded yet.

AshishDhama commented:
@hmhealey I have updated the branch

AshishDhama commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/104289

Willyfrog commented:
@koox00 friendly reminder to review :)

koox00 commented on packages/mattermost-redux/src/reducers/entities/channels.ts:
```diff
@@ -182,7 +182,7 @@ function channels(state: IDMappedObjects<Channel> = {}, action: GenericAction) {
         };
```
nvm I just saw @hmhealey's comment, this should be added in the new ticket to replace calls to fetch current teams channels and members with the new api to fetch all teams' channels.

esethna commented:
@hmhealey if you can please help with review asap as we are trying to merge before feature complete and growth spike

mattermod commented:
Triggering e2e testing with options:
```<nil>```

jgilliam17 commented:
/update-branch

jgilliam17 commented:
Thank you @AshishDhama 
Seeing 2 small theming/alignment issues, otherwise looks and works good.
I will check the  E2E report when the run completes and post here if there are any PR related failures.
1. Bot tag font theme seems off, only for multi-team user.
<img width="610" alt="Screen Shot 2022-01-12 at 10 54 26 PM" src="https://user-images.githubusercontent.com/52937121/149263973-4d4e68fd-8f37-4310-9717-9a46fd7d4c82.png">
<img width="630" alt="Screen Shot 2022-01-12 at 10 53 40 PM" src="https://user-images.githubusercontent.com/52937121/149263976-3e7de734-b566-4269-bf08-83a4a17c6d4c.png">
2. Mention badge may need to drop down for a few pixels to be vertically aligned. See above image.

michelengelen commented on actions/views/channel.js:
```diff
@@ -87,7 +93,13 @@ export function switchToChannelById(channelId) {
     return async (dispatch, getState) => {
         const state = getState();
-        const teamUrl = getCurrentRelativeTeamUrl(state);
+        let teamUrl;
+        const selectedTeamId = channel.team_id;
+        if (selectedTeamId) {
+            teamUrl = `/${getTeam(state, selectedTeamId).name}`;
+        } else {
+            teamUrl = getCurrentRelativeTeamUrl(state);
+        }
```
```suggestion
        const teamUrl = channel.team_id ? `/${getTeam(state, channel.team_id).name}` : getCurrentRelativeTeamUrl(state);
```

michelengelen commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -188,6 +194,10 @@ class SwitchChannelSuggestion extends Suggestion {
                 />
             );
         }
+        let teamName = null;
```
```suggestion

        let teamName = null;
```

michelengelen commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -188,6 +194,10 @@ class SwitchChannelSuggestion extends Suggestion {
         }
+        let teamName = null;
+        if (channel.team_id && team) {
+            teamName = (<span className='ml-2 suggestion-list__team-name'>{team.display_name}</span>);
```
```suggestion
            teamName = <span className='ml-2 suggestion-list__team-name'>{team.display_name}</span>;
```

michelengelen commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -204,16 +214,20 @@ class SwitchChannelSuggestion extends Suggestion {
-                    <span className='ml-2'>{description}</span>
+                <div className={classNames('suggestion-list__ellipsis', {'suggestion-list__flex': !isPartOfOnlyOneTeam})}>
+                    <span className='suggestion-list__main'>{name}</span>
+                    {isPartOfOnlyOneTeam && <span className='ml-2'>{description}</span>}
+                    {!isPartOfOnlyOneTeam &&
+                    <>
+                        {customStatus}
+                        {sharedIcon}
+                        {tag}
+                        {badge}
+                        {teamName}
+                    </>
+                    }
```
This can be simplified with writing a ternary instead of 2 different short-fuses:

```suggestion
                    {isPartOfOnlyOneTeam ? <span className='ml-2'>{description}</span> : (
                      <>
                          {customStatus}
                          {sharedIcon}
                          {tag}
                          {badge}
                          {teamName}
                      </>
                    )}
```

michelengelen commented on packages/mattermost-redux/src/actions/channels.ts:
```diff
@@ -539,6 +539,40 @@ export function fetchMyChannelsAndMembers(teamId: string): ActionFunc {
+        const state = getState();
+        const {currentUserId} = state.entities.users;
+        try {
+            const channelRequest = Client4.getAllTeamsChannels();
+            const channelsMembersRequest = Client4.getAllChannelsMembers(currentUserId);
+            channels = await channelRequest;
+            channelsMembers = await channelsMembersRequest;
```
What is the purpose of storing the Promises in variables here?
Wouldn't it work just as well when written like this?

```suggestion
            channels = await Client4.getAllTeamsChannels();
            channelsMembers = await Client4.getAllChannelsMembers(currentUserId);
```

michelengelen commented on packages/mattermost-redux/src/actions/channels.ts:
```diff
@@ -539,6 +539,40 @@ export function fetchMyChannelsAndMembers(teamId: string): ActionFunc {
 
+export function fetchAllMyTeamsChannelsAndChannelMembers(): ActionFunc {
+    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
+        let channels;
+        let channelsMembers;
+        const state = getState();
+        const {currentUserId} = state.entities.users;
```
nit: I would always prefer defining const values first

```suggestion
        const state = getState();
        const {currentUserId} = state.entities.users;
        let channels;
        let channelsMembers;
```

michelengelen commented on packages/mattermost-redux/src/selectors/entities/channels.ts:
```diff
@@ -336,6 +336,20 @@ export const getChannelSetInCurrentTeam: (state: GlobalState) => string[] = crea
     },
 );
 
+export const getChannelSetAllTeam: (state: GlobalState) => string[] = createSelector(
```
not entirely sure, but the naming here seems a bit off. Wouldn't this be better as `getChannelSetForWholeTeam`?

AshishDhama replied:
``getChannelSetForWholeTeam`` doesn't sound appropriate to me may be ``getChannelSetForAllTeam as whole is still giving the sense of single team.

michelengelen commented on packages/mattermost-redux/src/selectors/entities/channels.ts:
```diff
@@ -366,6 +380,21 @@ export const getChannelsInCurrentTeam: (state: GlobalState) => Channel[] = creat
+    getChannelSetAllTeam,
+    getCurrentUser,
+    (channels: IDMappedObjects<Channel>, getChannelSetAllTeam: string[], currentUser: UserProfile): Channel[] => {
+        let locale = General.DEFAULT_LOCALE;
+
+        if (currentUser && currentUser.locale) {
+            locale = currentUser.locale;
+        }
```
```suggestion
        const locale = currentUser?.locale || General.DEFAULT_LOCALE;
```

michelengelen reviewed (changes requested):
Looks good so far ... I just have some minor suggestions

jgilliam17 commented:
@AshishDhama here is the [E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/105154-e8a4c09-onprem-ent-webapp-pr-8888-mm-ee-test:e8a4c09/mochawesome.html)
Looks good overall, only two failed tests appear related to the PR changes 
MM-T1241
MM-T1225

jgilliam17 commented:
Thanks @AshishDhama 
The mention bag is vertically centered, but the theme is still off
This is what we currently have on master
<img width="603" alt="Screen Shot 2022-01-13 at 10 38 40 PM" src="https://user-images.githubusercontent.com/52937121/149447592-1ffaccf5-3eda-44ef-8371-e07e00dd7835.png">
<img width="598" alt="Screen Shot 2022-01-13 at 10 38 19 PM" src="https://user-images.githubusercontent.com/52937121/149447594-1d0c0489-f3a2-4256-8538-2be0e79228ec.png">

But on the PR test server theme of the bot tag and mention badge is like below

<img width="599" alt="Screen Shot 2022-01-13 at 10 40 40 PM" src="https://user-images.githubusercontent.com/52937121/149447711-1b8f5c3a-17ac-4281-a5ac-dc35602d45e8.png">
<img width="602" alt="Screen Shot 2022-01-13 at 10 40 28 PM" src="https://user-images.githubusercontent.com/52937121/149447712-a2e615dc-aab3-45eb-a742-94b77ccebb8b.png">

jgilliam17 commented:
/update-branch

jgilliam17 commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/107264

jgilliam17 commented:
@esethna @matthewbirtch 
 If nickname or full name is not set - username displays.
 If nickname and/or full name is set - username is missing. Is this as designed ?
<img width="1292" alt="Screen Shot 2022-01-21 at 12 06 24 PM" src="https://user-images.githubusercontent.com/52937121/150569849-46d53ce9-6f1f-4bc4-8347-6f0b12bc04e3.png">

Currently, we show username on daily, regardless 
<img width="602" alt="Screen Shot 2022-01-21 at 1 23 01 PM" src="https://user-images.githubusercontent.com/52937121/150580119-215d0452-2bc0-47e1-ba33-97fccda1b05d.png">

matthewbirtch commented:
> If nickname and/or full name is set - username is missing. Is this as designed ?

I think username should still be there — assuming we have the room. I think we may need to truncate though if the combination of display name, username and team name cause things to collide. 

@AshishDhama I was just playing with keeping the username, and I think there should still be enough space to keep it. Whoever did the original design work must have intentionally removed it but I'm thinking we should keep it if we think there is still space

jgilliam17 commented:
@AshishDhama one small suggestion 
We truncate to soon, there is more room on the modal since for DMs we do not display the team name 
<img width="603" alt="Screen Shot 2022-01-21 at 3 08 35 PM" src="https://user-images.githubusercontent.com/52937121/150593506-750a2555-58b6-4b73-998a-bf61dff995b9.png">

AshishDhama commented:
> @AshishDhama one small suggestion We truncate to soon, there is more room on the modal since for DMs we do not display the team name <img alt="Screen Shot 2022-01-21 at 3 08 35 PM" width="603" src="https://user-images.githubusercontent.com/52937121/150593506-750a2555-58b6-4b73-998a-bf61dff995b9.png">

@jgilliam17 if we add a guest tag or bot tag, custom status, and mention tag then it will almost make up for the space.

jgilliam17 commented:
Ah, I was editing my above post after I thought about other items that may display.

jgilliam17 commented:
Thanks for a reminder @AshishDhama Custom status added 
<img width="603" alt="Screen Shot 2022-01-21 at 3 19 09 PM" src="https://user-images.githubusercontent.com/52937121/150594749-740692b7-1cca-42d4-8e95-c95da0481c94.png">

matthewbirtch commented:
> There is still a little bit of room. I downgraded longname user to guest. This is what it looks like. @matthewbirtch thoughts?
> 
> <img alt="Screen Shot 2022-01-21 at 3 15 31 PM" width="600" src="https://user-images.githubusercontent.com/52937121/150594304-c459521b-0897-47aa-9cad-deaa4da031d6.png">

I'm fine with the truncation as is. By the time you add all the possible additions, there's not much to play with.

AshishDhama commented:
@michelengelen could please review it again.

michelengelen commented on packages/mattermost-redux/src/selectors/entities/channels.ts:
```diff
@@ -336,6 +336,20 @@ export const getChannelSetInCurrentTeam: (state: GlobalState) => string[] = crea
     },
 );
 
+export const getChannelSetForAllTeam: (state: GlobalState) => string[] = createSelector(
```
nit: is the naming correct? It seems as it it should be `getChannelSetForAllTeams`, or am I misreading something?

michelengelen commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -221,21 +218,18 @@ class SwitchChannelSuggestion extends Suggestion {
-                        }
+                        {(isPartOfOnlyOneTeam || channel.type === Constants.DM_CHANNEL) &&
+                            (<span className='ml-2 suggestion-list__desc'>{description}</span>
+                            )}
```
```suggestion
                        )}
```

AshishDhama replied:
Only curly can go down as ``)`` is part of ``span`` code snippet my editor doesn’t allow this indentation.

michelengelen commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -86,9 +85,7 @@ class SwitchChannelSuggestion extends Suggestion {
+                    <span className='badge'>
                         {unreadMentions}
                     </span>
                 </div>);
```
please always follow this convention of having the component inside the brackets on new lines and indented. This makes it way easier to read.

I know that we do not have a eslint rule for that, but this is done all over the code-base in this manor.

michelengelen reviewed (changes requested):
still 2 changes that need to be done before we can move on ... sry 🙇🏼

michelengelen commented on components/suggestion/switch_channel_provider.jsx:
```diff
@@ -221,21 +218,18 @@ class SwitchChannelSuggestion extends Suggestion {
-                        }
+                        {(isPartOfOnlyOneTeam || channel.type === Constants.DM_CHANNEL) &&
+                            (<span className='ml-2 suggestion-list__desc'>{description}</span>
+                            )}
```
this is the right way to write that in our context. How your IDE does not allow that should be of no concern.

```suggestion
                        {(isPartOfOnlyOneTeam || channel.type === Constants.DM_CHANNEL) && (
                            <span className='ml-2 suggestion-list__desc'>{description}</span>
                        )}
```

jgilliam17 reviewed (approved):
Thanks @AshishDhama 
Tested, looks good to merge.
- Verified Find Channels modal shows channels and users across all teams. Tested on server with database and additionally with Bleve search enabled. Testing using elasticsearch will be done after merge.

[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/107264-5ae060a-onprem-ent-webapp-pr-8888-mattermost/mm-ee-test5ae060a/mochawesome.html) looks good, no PR related failures

esethna commented:
@AshishDhama looks like reviews are complete and this can be merged?

amyblais commented:
@AshishDhama Can you help clarify the release note "API: modification in existing API endpoints to meet the Requirements (added new params)." - which API endpoints were modified and why?

AshishDhama commented:
@amyblais we have changed following end point ``${baseUrl}/api/v4/channels/search?system_console=false`` we have added additional param to fetch search results across teams.
and have added 
1. ``${baseUrl}/api/v4/users/me/channels`` (for fetching all the channel across teams)
2. ``${baseUrl}/api/v4/users/${userId/channel_members`` (for fetching all the channel_members across teams)
@agnivade please correct me If I am wrong.

agnivade commented:
Yep, those are the ones. The changes are documented in more details here: https://github.com/mattermost/mattermost-api-reference/commit/0c6827642dc286e25c2c9e4c9edcb1b53280b434.

cwarnermm commented:
Documented via https://github.com/mattermost/docs/pull/5391
